### ISSUE TITLE
wechaty-puppet-mock started as wechaty-puppet-wechat4u?

### ISSUE THREAD

**Issue by suntong** [wechaty/wechaty-getting-started#106]
## 0. Report Issue

1. Done running the following command
```
rm -rf package-lock.json
rm -rf node_modules
npm install
```

## 1. Versions

```
$ node -v
v10.20.0

$ npm ls wechaty wechaty-puppet wechaty-puppet-mock wechaty-puppet-puppeteer wechaty-puppet-padplus
wechaty-getting-started@0.1.20 /.../nodejs/wechaty-examples
├─┬ wechaty@0.38.4 
│ └── wechaty-puppet@0.24.0  deduped
├── wechaty-puppet@0.24.0 
├── wechaty-puppet-mock@0.20.4 
└── wechaty-puppet-puppeteer@0.18.6 

```

- What os are you using

```
$ lsb_release -a 
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.2 LTS
Release:        18.04
Codename:       bionic
```

## 2. Describe the bug

Try to start with `wechaty-puppet-mock` but what is actually started is `wechaty-puppet-wechat4u`.

## Anticipated Result

`wechaty-puppet-mock` kicks in and start sending messages, but get blocked with qrcode scan instead. 

## To Reproduce

1. run `WECHATY_LOG=silly WECHATY_PUPPET=wechaty-puppet-mock node examples/ding-dong-bot.js`

## 6. Full Output Logs

<details>
<summary>
Show Logs
</summary>

```shell
$ WECHATY_LOG=silly WECHATY_PUPPET=wechaty-puppet-mock node examples/ding-dong-bot.js 
16:06:07 SILL Config WECHATY_LOG set level to silly
16:06:07 INFO Config registering process.on("unhandledRejection") for development/debug
16:06:07 VERB Config constructor()
16:06:07 VERB Wechaty contructor()
16:06:07 VERB StateSwitch constructor(Wechaty, "{"log":{"enableTimestamp":true,"logLevel":5,"prefixFilter":{}}}")
16:06:07 VERB StateSwitch constructor(WechatyReady, "{"log":{"enableTimestamp":true,"logLevel":5,"prefixFilter":{}}}")
16:06:07 VERB Wechaty on(scan, function) registered
16:06:07 VERB Wechaty addListenerFunction(scan)
16:06:07 VERB Wechaty on(login, function) registered
16:06:07 VERB Wechaty addListenerFunction(login)
16:06:07 VERB Wechaty on(logout, function) registered
16:06:07 VERB Wechaty addListenerFunction(logout)
16:06:07 VERB Wechaty on(message, function) registered
16:06:07 VERB Wechaty addListenerFunction(message)
16:06:07 INFO Wechaty <wechaty-puppet-wechat4u>(ding-dong-bot) start() v0.38.4 is starting...
16:06:07 VERB Wechaty id: ck8xhai2l000055fu7fve9sqi
16:06:07 SILL StateSwitch <Wechaty> on() is false
16:06:07 SILL StateSwitch <WechatyReady> off() is true
16:06:07 VERB StateSwitch <WechatyReady> off(true) <- (true)
16:06:07 SILL StateSwitch <Wechaty> on() is false
16:06:07 VERB StateSwitch <Wechaty> on(pending) <- (false)
16:06:07 VERB MemoryCard constructor("ding-dong-bot")
16:06:07 VERB MemoryCard getStorage() for storage type: N/A
16:06:07 VERB getStorage name: ding-dong-bot, options: {"type":"file"}
16:06:07 VERB StorageFile constructor(ding-dong-bot, ...)
16:06:07 VERB StorageBackend constructor(ding-dong-bot, { type: file })
16:06:07 VERB MemoryCard load() from storage: StorageFile</lfs/workv/repo/gitwork/nodejs/wechaty-getting-started/ding-dong-bot.memory-card.json>
16:06:07 VERB StorageFile load() from /lfs/workv/repo/gitwork/nodejs/wechaty-getting-started/ding-dong-bot.memory-card.json
16:06:07 VERB MemoryCard load() file not exist, NOOP
16:06:07 VERB Wechaty initPuppet() wechaty-puppet-wechat4u
16:06:07 VERB MemoryCard multiplex(puppet)
16:06:07 VERB MemoryCard static multiplex(MemoryCard<ding-dong-bot>, puppet)
16:06:07 VERB MemoryCard constructor({"name":"ding-dong-bot","multiplex":{"name":"puppet","parent":{"options":{"name":"ding-dong-bot"},"name":"ding-dong-bot","payload":{},"multiplexNameList":[],"storage":{"name":"ding-dong-bot","options":{"type":"file"},"absFileName":"/lfs/workv/repo/gitwork/nodejs/wechaty-getting-started/ding-dong-bot.memory-card.json"}}}})
16:06:07 VERB PuppetManager resolve({puppet: wechaty-puppet-wechat4u, puppetOptions: undefined})
16:06:07 VERB PuppetManager resolveName(wechaty-puppet-wechat4u)
16:06:07 VERB PuppetManager checkModule(wechaty-puppet-wechat4u)
16:06:07 SILL PuppetManager checkModule() wechaty-puppet-wechat4u installed version 0.17.4 satisfied range ^0.17.4
16:06:07 VERB Puppet constructor({}) #0
16:06:07 VERB StateSwitch constructor(PuppetWechat4u, "{"log":{"enableTimestamp":true,"logLevel":5,"prefixFilter":{}}}")
16:06:07 VERB MemoryCard constructor(undefined)
16:06:07 VERB MemoryCard getStorage() for storage type: N/A
16:06:07 VERB MemoryCard load() from storage: N/A
16:06:07 VERB MemoryCard load() no storage
16:06:07 VERB Puppet constructor() watchdog timeout set to 60 seconds
16:06:07 VERB HotImport callerResolve(., /lfs/workv/repo/gitwork/nodejs/wechaty-getting-started/node_modules/wechaty-puppet/dist/src/puppet.js)
16:06:07 SILL HotImport callerResolve() callsites() file=/lfs/workv/repo/gitwork/nodejs/wechaty-getting-started/node_modules/hot-import/dist/src/hot-import.js, type=Object
16:06:07 SILL HotImport callerResolve() callsites() file=/lfs/workv/repo/gitwork/nodejs/wechaty-getting-started/node_modules/wechaty-puppet/dist/src/puppet.js, type=null
16:06:07 SILL HotImport callerResolve() callsites() file=/lfs/workv/repo/gitwork/nodejs/wechaty-getting-started/node_modules/wechaty/node_modules/wechaty-puppet-wechat4u/dist/src/puppet-wechat4u.js, type=null
16:06:07 SILL HotImport callerResolve() callerFile=/lfs/workv/repo/gitwork/nodejs/wechaty-getting-started/node_modules/wechaty/node_modules/wechaty-puppet-wechat4u/dist/src/puppet-wechat4u.js
16:06:07 VERB Puppet constructor() childClassPath=/lfs/workv/repo/gitwork/nodejs/wechaty-getting-started/node_modules/wechaty/node_modules/wechaty-puppet-wechat4u/dist/src
16:06:07 VERB Puppet constructor() memory.load() done
16:06:07 VERB Puppet setMemory()
16:06:07 VERB Wechaty initPuppetEventBridge(Puppet#0<PuppetWechat4u>(ding-dong-bot))
16:06:07 VERB Wechaty initPuppetEventBridge() puppet.on(friendship) (listenerCount:0) registering...
16:06:07 VERB Wechaty initPuppetEventBridge() puppet.on(login) (listenerCount:0) registering...
16:06:07 VERB Wechaty initPuppetEventBridge() puppet.on(logout) (listenerCount:0) registering...
16:06:07 VERB Wechaty initPuppetEventBridge() puppet.on(message) (listenerCount:0) registering...
16:06:07 VERB Wechaty initPuppetEventBridge() puppet.on(room-invite) (listenerCount:0) registering...
16:06:07 VERB Wechaty initPuppetEventBridge() puppet.on(room-join) (listenerCount:0) registering...
16:06:07 VERB Wechaty initPuppetEventBridge() puppet.on(room-leave) (listenerCount:0) registering...
16:06:07 VERB Wechaty initPuppetEventBridge() puppet.on(room-topic) (listenerCount:0) registering...
16:06:07 VERB Wechaty initPuppetEventBridge() puppet.on(scan) (listenerCount:0) registering...
16:06:07 VERB Wechaty initPuppetEventBridge() puppet.on(dong) (listenerCount:0) registering...
16:06:07 VERB Wechaty initPuppetEventBridge() puppet.on(error) (listenerCount:0) registering...
16:06:07 VERB Wechaty initPuppetEventBridge() puppet.on(heartbeat) (listenerCount:1) registering...
16:06:07 VERB Wechaty initPuppetEventBridge() puppet.on(ready) (listenerCount:0) registering...
16:06:07 VERB Wechaty initPuppetEventBridge() puppet.on(reset) (listenerCount:1) registering...
16:06:07 VERB Wechaty initAccessory(Puppet#0<PuppetWechat4u>(ding-dong-bot))
16:06:07 SILL Accessory <Contact> static set wechaty = "Wechaty#ck8xhai2l000055fu7fve9sqi<wechaty-puppet-wechat4u>(ding-dong-bot)"
16:06:07 SILL Accessory <ContactSelf> static set wechaty = "Wechaty#ck8xhai2l000055fu7fve9sqi<wechaty-puppet-wechat4u>(ding-dong-bot)"
16:06:07 SILL Accessory <Friendship> static set wechaty = "Wechaty#ck8xhai2l000055fu7fve9sqi<wechaty-puppet-wechat4u>(ding-dong-bot)"
16:06:07 SILL Accessory <Image> static set wechaty = "Wechaty#ck8xhai2l000055fu7fve9sqi<wechaty-puppet-wechat4u>(ding-dong-bot)"
16:06:07 SILL Accessory <Message> static set wechaty = "Wechaty#ck8xhai2l000055fu7fve9sqi<wechaty-puppet-wechat4u>(ding-dong-bot)"
16:06:07 SILL Accessory <Room> static set wechaty = "Wechaty#ck8xhai2l000055fu7fve9sqi<wechaty-puppet-wechat4u>(ding-dong-bot)"
16:06:07 SILL Accessory <RoomInvitation> static set wechaty = "Wechaty#ck8xhai2l000055fu7fve9sqi<wechaty-puppet-wechat4u>(ding-dong-bot)"
16:06:07 SILL Accessory <Tag> static set wechaty = "Wechaty#ck8xhai2l000055fu7fve9sqi<wechaty-puppet-wechat4u>(ding-dong-bot)"
16:06:07 SILL Accessory <Contact> static set puppet = "Puppet#0<PuppetWechat4u>(ding-dong-bot)"
16:06:07 SILL Accessory <ContactSelf> static set puppet = "Puppet#0<PuppetWechat4u>(ding-dong-bot)"
16:06:07 SILL Accessory <Friendship> static set puppet = "Puppet#0<PuppetWechat4u>(ding-dong-bot)"
16:06:07 SILL Accessory <Image> static set puppet = "Puppet#0<PuppetWechat4u>(ding-dong-bot)"
16:06:07 SILL Accessory <Message> static set puppet = "Puppet#0<PuppetWechat4u>(ding-dong-bot)"
16:06:07 SILL Accessory <Room> static set puppet = "Puppet#0<PuppetWechat4u>(ding-dong-bot)"
16:06:07 SILL Accessory <RoomInvitation> static set puppet = "Puppet#0<PuppetWechat4u>(ding-dong-bot)"
16:06:07 SILL Accessory <Tag> static set puppet = "Puppet#0<PuppetWechat4u>(ding-dong-bot)"
16:06:07 SILL StateSwitch <PuppetWechat4u> on() is false
16:06:07 VERB StateSwitch <PuppetWechat4u> on(pending) <- (false)
16:06:07 VERB MemoryCard <puppet> get(PUPPET-WECHAT4U)
16:06:07 SILL StateSwitch <PuppetWechat4u> on() is pending
16:06:07 VERB StateSwitch <PuppetWechat4u> on(true) <- (pending)
16:06:07 VERB Wechaty on(heartbeat, function) registered
16:06:07 VERB Wechaty addListenerFunction(heartbeat)
16:06:07 SILL StateSwitch <Wechaty> on() is pending
16:06:07 VERB StateSwitch <Wechaty> on(true) <- (pending)
16:06:07 INFO StarterBot Starter Bot Started.
▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄
█ ▄▄▄▄▄ █▄▀ ▀  █▄▄▀█▄▄█ ▄▄▄▄▄ █
█ █   █ █   █▀▄   ▀ ▄▀█ █   █ █
█ █▄▄▄█ █▄█▀ ▄  ▄▄▀▀ ██ █▄▄▄█ █
█▄▄▄▄▄▄▄█▄█ █ █ █ ▀▄▀▄█▄▄▄▄▄▄▄█
█   ▄▀▀▄▀▄█ ▄ █ ▄▄▄ ▀█ ▀▀ ▄▄█ █
██▄▀▀██▄█▀  ▄▄▀ █▄█▀█ ▄▄█▀█▄▄▀█
██▀ ▀█ ▄▀▄ ▄▀ ▄  ▀▄█ █▄▀██▄ ▀▀█
█▀█▀█▀▄▄ ▄▀▄ ██ ▀▀▄▄▄▄ ▀  █▄ ▄█
█▄▄ ▀▀▀▄▀▄▀██▄▄ ▀ ▀▄▄  █ ▄ ▀ ██
█▄█▄▄█▀▄ █ █▀██▄▄▄▀▀█▀▀▀▄▀█ ███
██▄█▄█▄▄█▀  ██▀▀█  ▀▀ ▄▄▄   ▀▀█
█ ▄▄▄▄▄ ██▀▀▀▀▀▀█ █ ▄ █▄█  █▄██
█ █   █ █▀█▄ █▀▀█ ▄█▀  ▄  ▀▄▄ █
█ █▄▄▄█ █ ▄ █ ▀▄█▀█ ▄ ██▄▀▀▄▀▄█
█▄▄▄▄▄▄▄█▄▄▄▄███▄███▄▄▄██▄██▄██

16:06:09 INFO StarterBot onScan: Waiting(2) - https://api.qrserver.com/v1/create-qr-code/?data=https%3A%2F%2Flogin.weixin.qq.com%2Fl%2FYa5e-rOLqg%3D%3D
16:06:09 SILL StateSwitch <PuppetWechat4u> off() is false
16:06:35 SILL StateSwitch <PuppetWechat4u> off() is false

```

</details>

[bug]

**Comment by suntong** [comment]
Can we remove the line

https://github.com/wechaty/wechaty-getting-started/blob/master/examples/ding-dong-bot.js#L59

so that it can be started as any puppet please?

**Comment by suntong** [comment]
Ops, no, tried that and got the following:

```
odules/wechaty/dist/src/user/message.js:682:16)
    at Wechaty.<anonymous> (/path/to/nodejs/wechaty-getting-started/node_modules/wechaty/dist/src/wechaty.js:470:35)
    at Generator.next (<anonymous>)
    at /path/to/nodejs/wechaty-getting-started/node_modules/wechaty/dist/src/wechaty.js:8:71
    at new Promise (<anonymous>)
    at __awaiter (/path/to/nodejs/wechaty-getting-started/node_modules/wechaty/dist/src/wechaty.js:4:12)
    at PuppetMock.puppet.on (/path/to/nodejs/wechaty-getting-started/node_modules/wechaty/dist/src/wechaty.js:468:55)
    at PuppetMock.emit (events.js:198:13)
    at PuppetMock.EventEmitter.emit (domain.js:448:20)
    at PuppetMock.emit (/path/to/nodejs/node_modules/wechaty-puppet/dist/src/puppet.js:163:22)
    at Timeout.loopTimer.setInterval [as _onTimeout] (/path/to/nodejs/node_modules/wechaty-puppet-mock/dist/src/puppet-mock.js:63:22)
    at ontimeout (timers.js:436:11)
    at tryOnTimeout (timers.js:300:5)
    at listOnTimeout (timers.js:263:5)
    at Timer.processTimers (timers.js:223:10)
(node:6602) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 24)
^C

```

Wechaty framework used to have the power to change the puppet on the fly:

```
WECHATY_PUPPET=wechaty-puppet-mock node starter-bot.js
WECHATY_PUPPET=wechaty-puppet-puppeteer node starter-bot.js
```

I.e., the same script, `starter-bot.js`, does not need to specify which puppet it is using, all can be provided at run time from command line. 

But not any more.

**Comment by suntong** [comment]
problem fixed now, pending #107 to close this.
